src/api/adapter_manager.rs

```diff
@@ -22,33 +27,15 @@ where
     PeripheralType: Peripheral,
 {
     peripherals: Arc<DashMap<BDAddr, PeripheralType>>,
-
-    // Sender is never handled mutably, but mpsc's Sender is Send only, not Sync,
-    // so we can't just wrap it in an Arc to pass it around as part of the adapter
-    // struct. I also don't want to use crossbeam channels, to avoid type leakage.
-    //
-    // This will be fixed when we go async in 1.0 and can use stream traits. For
-    // now, we deal with the lock timing.
-    event_sender: Arc<Mutex<Sender<CentralEvent>>>,
-
-    // Normally we'd just return the event receiver when an adapter is created.
-    // However, since adapters are cloned and retrieved via lists, this is really
-    // hard without changing the fundamentals of the API (which I want to do at
-    // some point, but not now). Storing an option here means that we'll only ever
-    // have one event receiver (as mpsc isn't clonable, which is what we want on
-    // the receiver side anyways), but means we also don't have to deal with the
-    // adapter API yet.
-    event_receiver: Arc<Mutex<Option<Receiver<CentralEvent>>>>,
+    async_senders: Arc<Mutex<Vec<UnboundedSender<CentralEvent>>>>,
```

Comment:
Can we just make this a tokio Broadcast channel?

Reply:
Unfortunately `tokio_stream::wrappers::BroadcastStream<T>` implements `Stream<Item=Result<T, BroadcastStreamRecvError>>` rather than `Stream<Item=T>`, which would pollute our API and mean clients have to handle `BroadcastStreamRecvError`. It would also mean adding a dependency on `tokio_stream`, though that's less of an issue.

Reply:
No, it won't. You use tokio_stream to rewrap the Stream, I do this in Buttplug already with no API pollution.

https://github.com/buttplugio/buttplug-rs/blob/master/buttplug/src/util/stream.rs#L5

Reply:
It looks like that stream will end if `Receiver::recv` returns `Err(RecvError::Lagged)`, which seems wrong.